name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/views/SecuritiesChart.java

```diff
@@ -1246,59 +1255,50 @@ private void addDividendMarkerLines(ChartInterval chartInterval)
             if (chartConfig.contains(ChartDetails.SHOW_DATA_LABELS))
             {
                 customPaintListeners.add(event -> {
+                    FontMetrics fontMetrics = event.gc.getFontMetrics();
+
+                    StringBuilder labelBuilder = new StringBuilder();
+
+                    // Three levels of the label
+                    int[] yPosLabels = new int[3];
+                    int[] lastWriteLabels = new int[3];
+                    Arrays.fill(lastWriteLabels, 0);
+
+                    int symbolSize = border.getSymbolSize();
+                    int labelHeight = fontMetrics.getHeight();
+                    int halfLabelHeight = labelHeight / 2;
+
                     IAxis xAxis = chart.getAxisSet().getXAxis(0);
                     IAxis yAxis = chart.getAxisSet().getYAxis(0);
 
-                    int yPosLabel = 0;
-                    int lastWriteLabelLevel1 = 0;
-                    int lastWriteLabelLevel2 = 0;
-                    int lastWriteLabelLevel3 = 0;
-
                     for (int index = 0; index < dates.length; index++)
                     {
-                        boolean freeSpaceForLabelLevel1 = true;
-                        boolean freeSpaceForLabelLevel2 = true;
-                        boolean freeSpaceForLabelLevel3 = true;
-
                         int x = xAxis.getPixelCoordinate(dates[index].getTime());
                         int y = yAxis.getPixelCoordinate(values[index]);
 
-                        String label = getDividendLabel(dividends.get(index));
-                        Point textExtent = event.gc.textExtent(label);
+                        labelBuilder.setLength(0);
+                        labelBuilder.append(getDividendLabel(dividends.get(index)));
 
-                        event.gc.setForeground(Colors.theme().defaultForeground());
+                        int labelWidth = event.gc.stringExtent(labelBuilder.toString()).x;
+                        int halfLabelWidth = labelWidth / 2;
 
-                        if (((x - (textExtent.x / 2)) - lastWriteLabelLevel1) <= 0)
-                            freeSpaceForLabelLevel1 = false;
+                        boolean labelPlaced = false;
 
-                        if (((x - (textExtent.x / 2)) - lastWriteLabelLevel2) <= 0)
-                            freeSpaceForLabelLevel2 = false;
-
-                        if (((x - (textExtent.x / 2)) - lastWriteLabelLevel3) <= 0)
-                            freeSpaceForLabelLevel3 = false;
-
-                        if (freeSpaceForLabelLevel1 || freeSpaceForLabelLevel2 || freeSpaceForLabelLevel3)
+                        for (int level = 0; level < 3 && !labelPlaced; level++)
```

Comment:
Instead of the boolean labelPlaced, I think you can just call "break" and you will jump out of the loop.